# Logistic Regression Classification

**Status:** Planned

Work through the sections in order. This notebook uses scikit-learn's built-in breast-cancer dataset, so no download is needed. Keep the test set untouched until the final evaluation.

Related note: [`ai-docs/docs/ml/model-evaluation.md`](../../ai-docs/docs/ml/model-evaluation.md)

## 1. Setup

**Question:** Which tools will you use for loading data, splitting it, scaling features, fitting logistic regression, and evaluating a classifier?

In [ ]:
# TODO: import numpy, pandas, matplotlib.pyplot, and the scikit-learn tools below.
# Suggested tools: load_breast_cancer, train_test_split, StandardScaler,
# Pipeline, LogisticRegression, accuracy_score, precision_score, recall_score,
# f1_score, confusion_matrix, ConfusionMatrixDisplay, roc_auc_score
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix,
    ConfusionMatrixDisplay,
    roc_auc_score,
)



ImportError: cannot import name 'train_test_split' from 'sklearn.datasets' (/home/nirupam/Projects/ai-learning/ai-engineering-notes/.venv/lib/python3.13/site-packages/sklearn/datasets/__init__.py)

## 2. Load and inspect the data

**Questions:**

1. How many examples and feature columns are present?
2. What are the two target classes?
3. Is this a regression or classification task, and why?

In [8]:
# TODO: load the dataset, create a DataFrame named df, and add the target column.
# Inspect df.shape, the class names, and target.value_counts().

breast_cancer_data = load_breast_cancer()
df = pd.DataFrame(breast_cancer_data.data, columns=breast_cancer_data.feature_names)
df['target'] = breast_cancer_data.target

print(df.shape)
print(df.columns)
print(breast_cancer_data.target_names)
print(df['target'].value_counts())
df.head()


(569, 31)
Index(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error', 'fractal dimension error',
       'worst radius', 'worst texture', 'worst perimeter', 'worst area',
       'worst smoothness', 'worst compactness', 'worst concavity',
       'worst concave points', 'worst symmetry', 'worst fractal dimension',
       'target'],
      dtype='str')
['malignant' 'benign']
target
1    357
0    212
Name: count, dtype: int64


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,0


Write your answers here:

1. 569 examples and 31 features are present
2. two target classes are malignant and benign
3. This is a classification task as we need to determine between two classes which is likely

## 3. Create a holdout test set

Use an 80/20 split with `random_state=42` and `stratify=y`. Stratification preserves the class balance in both sets.

**Question:** Why is stratification useful when the target classes are not equally common?

In [ ]:
# TODO: create X and y, then split them into X_train, X_test, y_train, and y_test.
# Print their shapes and the normalized class proportions for y_train and y_test.

X_train, X_test, y_train, y_test = train_test_split(
    df.drop(columns='target'),
    df['target'],
    test_size=0.2,
    random_state=42,
    stratify=df['target']
)

Your answer: 

## 4. Establish a majority-class baseline

Make a baseline that predicts the most common training class for every test example.

**Question:** Why can accuracy be misleading when one class is much more common than the other?

In [ ]:
# TODO: find the most common class in y_train and predict it for every y_test example.
# Print the baseline accuracy and its confusion matrix.



Your answer: 

## 5. Train logistic regression

Create a `Pipeline` containing `StandardScaler` and `LogisticRegression(max_iter=1000)`. The scaler must be inside the pipeline so it is fit only on the training data.

**Question:** What does the model's predicted probability represent?

In [ ]:
# TODO: build, fit, and use the pipeline.
# Store class predictions in y_pred and positive-class probabilities in y_prob.



Your answer: 

## 6. Evaluate the classifier

Calculate accuracy, precision, recall, F1, and ROC-AUC. Display a labelled confusion matrix.

**Questions:**

1. How does the logistic-regression model compare with the majority-class baseline?
2. What kinds of mistakes are false positives and false negatives for this dataset?
3. Which metric would become especially important if missing a positive case were costly?

In [ ]:
# TODO: calculate the metrics and display a confusion matrix.
# Hint: check dataset.target_names before deciding which integer is the positive class.



Write your answers here:

1. 
2. 
3. 

## 7. Explore the decision threshold

By default, `predict()` uses a probability threshold of `0.5`. Create predictions using a different threshold, such as `0.3` or `0.7`, then compare precision and recall.

**Question:** What trade-off do you observe when the threshold changes?

In [ ]:
# TODO: set a threshold, create threshold_predictions from y_prob, and compare precision and recall.



Your answer: 

## Challenge

Choose one extension:

- Plot an ROC curve and explain what ROC-AUC measures.
- Try three `C` values for logistic regression and compare validation results without touching the test set.
- Inspect the largest positive and negative coefficients after scaling, then explain why coefficients show association rather than causation.

In [ ]:
# TODO: implement one challenge.



## Reflection

- What did you learn about probabilities versus class labels?
- Which metric would you choose for a high-stakes use case, and why?
- What will you explore next?